src/box/iproto.cc

```diff
@@ -821,6 +927,51 @@ iproto_connection_input_buffer(struct iproto_connection *con)
 	return new_ibuf;
 }
 
+/**
+ * Checks is message belongs to stream (stream_id != 0), and if it
+ * is so creates new/stream or gets stream from connection streams
+ * hash table. Puts message to stream pending messages list.
+ * @retval 0 - if message ready to push
+ *         1 - if message puts in pending list
+ *        -1 - if error occurs
+ */
+static int
+iproto_set_msg_stream(struct iproto_msg *msg)
+{
+	uint32_t stream_id = msg->header.stream_id;
+	if (stream_id == 0)
+		return 0;
+
+	bool skip_push;
+	struct iproto_connection *con = msg->connection;
+	struct stream *stream = NULL;
+	mh_int_t pos = mh_streams_find(con->streams, stream_id, 0);
+	if (pos == mh_end(con->streams)) {
+		stream = iproto_stream_new(msg);
+		if (stream == NULL)
+			return -1;
+		pos = mh_streams_put(con->streams, (const struct stream **)
+				     &stream, NULL, 0);
+		if (pos == mh_end(con->streams)) {
+			iproto_stream_delete(stream);
+			diag_set(OutOfMemory, pos + 1, "mh_streams_put",
+				 "mh_streams_node");
+			return -1;
+		}
+	}
+#ifndef NDEBUG
```

Comment:
We can't do it in `iproto_msg_new` because not all messages belongs to stream. Also we call `iproto_msg_decode` after `iproto_message_new`. But exactly in `iproto_msg_decode` we determine whether the message belongs to the stream.